### PR TITLE
mmap: explicitly expose capability permissions

### PR DESCRIPTION
**Pull request by brooksdavis**
Introduce three new PROT_ values PROT_READ_CAP, PROT_WRITE_CAP, and PROT_NO_IMPLY_CAP.  They combine to allow capability permissions to be implied in unmodified code using PROT_READ and PROT_WRITE allowing capability read and write permissions to be set or unset explicity.

If any of PROT_READ_CAP, PROT_WRITE_CAP, or PROT_NO_IMPLY_CAP are set, then the values of the PROT_READ_CAP and PROT_WRITE_CAP flag bits define the page protections and capability permissions for a given mapping.
    
In the underlying implementation, PROT_READ_CAP and PROT_WRITE_CAP map to VM_PROT_READ_CAP and VM_PROT_WRITE_CAP respectively and PROT_NO_IMPLY_CAP maps to a new VM_PROT_NO_IMPLY_CAP. VM_PROT_NO_IMPLY_CAP used transiently in fo_mmap implementations to avoid accidently adding capability permission and is also added to vm_entry's max_protection to allow superset tests to succeed when reducing capability permissions on a mapping via mmap or mprotect.

This differs from the approach in #884 in that PROT_NO_IMPLY_CAP is carried down as VM_PROT_NO_IMPLY_CAP so that VM_PROT_ADD_CAP doesn't add capabilities it shouldn't. The one implementation thing that's a bit wonky is that you have to add VM_PROT_NO_IMPLY_CAP to protections before you starting doing subset comparisons. I've mostly solved this by adding VM_PROT_NO_IMPLY_CAP to max_protection in the vm entry (at that point you're definitely done implying capability permissions).

### CODE REVIEW

File: sys/vm/vm_mmap.c [hunk 1]
```diff
@@ -234,6 +234,35 @@ vm_wxcheck(struct proc *p, char *call)
...
+		SYSERRCAUSE("%s: PROT_READ_CAP without PROT_READ in %s",
+		    func, protname);
+		return (ENOTSUP);
+	}
+	if ((prot & (PROT_WRITE_CAP | PROT_WRITE)) == PROT_WRITE_CAP) {
+		SYSERRCAUSE("%s: PROT_WRITE_CAP without PROT_WRITE in %s",
+		    func, protname);
+		return (ENOTSUP);
+	}
+	/*
```
Comment (bsdjhb) [inline sys/vm/vm_mmap.c:250]: I think this detail might be worth calling out explicitly in the commit log.  Arguably if we want to go this route, then I would actually suggest that we don't even expose separate R/W bits via mmap().  That is, I would just have `PROT_CAP` and `PROT_NO_CAP` exposed to userspace.  Potentially we could collapse this down further in the VM to a single "VM_PROT_CAP" as well without separate R/W bits there.

File: bin/cheribsdtest/cheribsdtest_vm.c [hunk 2]
```diff
@@ -115,6 +115,91 @@ CHERIBSDTEST(vm_tag_mmap_anon,
...
+	/* Mixing implied and explict protections */
+	mmap_check_bad_protections(PROT_READ | PROT_READ_CAP |
+	    PROT_MAX(PROT_READ), ENOTSUP);
+
+	/* Disallowed explicit capability protection combinations */
+	mmap_check_bad_protections(PROT_READ_CAP, ENOTSUP);
+	mmap_check_bad_protections(PROT_WRITE_CAP, ENOTSUP);
+	mmap_check_bad_protections(PROT_MAX(PROT_READ_CAP), ENOTSUP);
+	mmap_check_bad_protections(PROT_MAX(PROT_WRITE_CAP), ENOTSUP);
+	mmap_check_bad_protections(PROT_READ | PROT_WRITE | PROT_READ_CAP,
```
[the remark this answers is not in the document]
Comment (brooksdavis) [reply bin/cheribsdtest/cheribsdtest_vm.c:191]: My general view is that it's of limited use (I can't think of any use cases) and both future PTE and capability permission encodings may render it impossible. By forbidding it now we ensure that no one relies on it as I'd prefer not to end up in a situation were have to fail open due to running code that tries to do this.

File: lib/libsys/mmap.2 [hunk 3]
```diff
@@ -113,8 +113,47 @@ Pages may be read.
 Pages may be written.
 .It Dv PROT_EXEC
 Pages may be executed.
+.It Dv PROT_READ_CAP
+CHERI capabilities may be read from pages.
+.It Dv PROT_WRITE_CAP
+CHERI capabilities may be written to pages.
+.It Dv PROT_NO_IMPLY_CAP
```
[the remark this answers is not in the document]
Comment (jrtc27) [reply lib/libsys/mmap.2:120]: Or PROT_CAP_NONE?

[the remark this answers is not in the document]
Comment (brooksdavis) [reply lib/libsys/mmap.2:120]: I've written PROT_NONE_CAP and PROT_NO_CAP in past (partial) attempts. I think I want to avoid a NONE variant because it's not a no-op like PROT_NONE. The way I put it in the manual is more like PROT_HONOR_OTHER_CAP_BITS, but that's probably too wordy.

File: bin/cheribsdtest/cheribsdtest_vm.c (hunk 2 above)
[the remark this answers is not in the document]
Comment (bsdjhb) [reply bin/cheribsdtest/cheribsdtest_vm.c:191]: In that case, I would argue for my last comment which is to just have a single `PROT_CAP`.

[the remark this answers is not in the document]
Comment (brooksdavis) [reply bin/cheribsdtest/cheribsdtest_vm.c]: I've added a commit (to be squashed before merge) which combines them. This is a little awkward PROT_ and VM_PROT flags no longer match, but it is simplifying.

File: lib/libsys/mmap.2 (hunk 3 above)
[the remark this answers is not in the document]
Comment (brooksdavis) [reply lib/libsys/mmap.2]: In my update I switched to PROT_NO_CAP.

File: bin/cheribsdtest/cheribsdtest_vm.c (hunk 2 above)
[the remark this answers is not in the document]
Comment (bsdjhb) [reply bin/cheribsdtest/cheribsdtest_vm.c]: I think if this is the direction we want, we might want to merge the VM_PROT_*CAP flags as well, but I'm not sure of that yet and how that interacts with revocation (e.g. might there still be a case where you want to do a COW-like thing where you only allow cap loads but not cap stores or vice versa).

File: lib/libsys/mmap.2 [hunk 4]
```diff
@@ -113,8 +113,45 @@ Pages may be read.
...
+.Dv PROT_WRITE
+respectively, unless the underlying backing store can not safety support
+capabilities (e.g., a
+.Dv MAP_SHARED
+mapping of a file).
+If either of
+.Dv PROT_CAP
+or
+.Dv PROT_NO_CAP
+are set, then capability permissions will not be implied.
```
Comment (bsdjhb) [inline lib/libsys/mmap.2:141]:
```suggestion
.Dv PROT_CAP
if either of
.Dv PROT_READ
or
.Dv PROT_WRITE
is set unless the underlying backing store can not safely support
capabilities (e.g., a
.Dv MAP_SHARED
mapping of a file).
If
.Dv PROT_NO_CAP
is set, then capability permissions will not be implied.
```

This is specific to squashing down to `PROT_CAP` as then I think the "respectively" language is no longer needed.  Also, I think it's kind of redundant (and possibly confusing) to say that an explicit `PROT_CAP` doesn't imply `PROT_CAP`.

File: lib/libsys/mmap.2 [hunk 5]
```diff
@@ -113,8 +113,45 @@ Pages may be read.
...
+is passed, at least one of
+.Dv PROT_READ
+and
+.Dv PROT_WRITE
+is required.
+On non-CHERI platforms the
+.Dv PROT_CAP
+and
+.Dv PROT_NO_CAP
+bits have no effect.
```
Comment (bsdjhb) [inline lib/libsys/mmap.2:153]:
```suggestion
flags have no effect.
```

File: lib/libsys/mmap.2 [hunk 6]
```diff
@@ -113,8 +113,45 @@ Pages may be read.
...
+.Pp
 In addition to these protection flags,
 .Fx
 provides the ability to set the maximum protection of a region allocated by
```
Comment (bsdjhb) [inline lib/libsys/mmap.2:157]: Do we want any manpage language for the fact that PROT_MAX() must include an explicit PROT*_CAP flag if the regular permissions flags include one of those flags?
